File: src/_3_polish/java/ArithmeticStack.java
package _3_polish.java;

import java.util.Deque;
import java.util.LinkedList;
import java.util.function.BiFunction;

public class ArithmeticStack {

    private final Deque<Double> stack = new LinkedList<Double>();

    public enum Operation {
        PLUS((x, y) -> x + y),
        MINUS((x, y) -> y - x),
        TIMES((x, y) -> x * y),
        DIV((x, y) -> y / x);

        final BiFunction<Double, Double, Double> function;

        Operation(BiFunction<Double, Double, Double> function) {
            this.function = function;
        }
    }

    public void push(double x) {
        stack.push(x);
    }

    public double top() {
        return stack.peek();
    }

    public void execute(Operation op) {
        stack.push(op.function.apply(stack.pop(), stack.pop()));
    }
}

File: src/_3_polish/kotlin/ArithmeticStack.kt
package _3_polish.kotlin

import java.util.*

class ArithmeticStack {

    // Type inference
    private val stack = LinkedList<Double>()

    // Enums and lambdas
    enum class Operation(internal val function: (Double, Double) -> Double) {
        PLUS({ x, y -> x + y }),
        MINUS({ x, y -> y - x }),
        TIMES({ x, y -> x * y }),
        DIV({ x, y -> y / x })
    }

    // Expression function syntax
    fun push(x: Double) = stack.push(x)

    fun top() = stack.peek()

    // Lambda invocation
    fun execute(op: Operation) = stack.push(op.function(stack.pop(), stack.pop()))
}